sklearn/ensemble/_gradient_boosting.pyx

```diff
@@ -84,22 +88,111 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
         shape ``(n_samples, K)``.
     """
     cdef Py_ssize_t i
-    cdef int32 node_id
     cdef Node *node
     for i in range(n_samples):
         node = root_node
         # While node not a leaf
-        while node.left_child != -1 and node.right_child != -1:
+        while node.left_child != TREE_LEAF:
             if X[i * n_features + node.feature] <= node.threshold:
                 node = root_node + node.left_child
             else:
                 node = root_node + node.right_child
         out[i * K + k] += scale * value[node - root_node]
 
+def _predict_regression_tree_stages_sparse(np.ndarray[object, ndim=2] estimators,
+                                           object X, double scale,
+                                           np.ndarray[float64, ndim=2] out):
+    """Predicts output for regression tree inplace and adds scaled value to ``out[i, k]``.
+
+    The function assumes that the ndarray that wraps ``X`` is csr_matrix.
+    """
+    cdef DTYPE_t* X_data = <DTYPE_t*>(
+        <np.ndarray[ndim=1, dtype=DTYPE_t]>X.data).data
+    cdef INT32_t* X_indices = <INT32_t*>(
+        <np.ndarray[ndim=1, dtype=INT32_t]> X.indices).data
+    cdef INT32_t* X_indptr = <INT32_t*>(
+        <np.ndarray[ndim=1, dtype=INT32_t]> X.indptr).data
```

Comment:
Actually I don't know :), I saw this some Cython/Numpy tutorials, and I think it's always better to specify everything explicitly, when we want maximum performance.